posts/frangeris/01.reusable-dynamic-modal-on-vue3.md

```diff
@@ -0,0 +1,215 @@
+---
+title: Reusable dynamic modal on Vue 3
+published: false
+description: Quick explanation on how to make use of new composition API on Vue 3 to get a reusable modal
+tags: "javascript, vue, webdev, tutorial"
+cover_image: ./assets/01.cover.png
+---
+
+## Reusable dynamic modal on Vue 3
+
+Most of the time on frontend development the best way to keep a consistent way of building components is to try to make them reusable every time we can, but sometimes the framework itself can make it a bit harder if we don’t have deep knowledge of its internal API, specifically the way it handle view instance and state component data.
+
+What makes a good reusable modal?
+
+- ✅ Dynamic views
+- ✅ Customizable actions with callbacks (buttons)
+- ✅ easy-peasy instantiation
+
+For simplicity we will be using:
+
+- [DaisyUI](https://daisyui.com) as component library ([modal](https://daisyui.com/components/modal/))
+- [Pinia](https://pinia.vuejs.org/) as state management vue store
+- [Vue 3](https://vuejs.org/) with the composition API [with typescript](https://vuejs.org/guide/typescript/overview.html)
+
+Let’s start by defining a very basic modal template using daisy classes:
+
+```html
+<!-- @/components/x-modal.vue -->
+
+<template>
+  <div>
+    <div class="modal modal-open">
+      <div class="modal-box relative">
+        <label class="btn btn-sm btn-circle absolute right-2 top-2">✕</label>
+        <h1>Hey, it works 👏🏽</h1>
+      </div>
+    </div>
+  </div>
+</template>
+```
+
+To be able to access it from anywhere and render any type of view, the modal needs to be accessible from any component that wants to use it, for this reason the modal component needs to be placed on the root parent view where all the views will be rendered, assuming [Vue router](https://router.vuejs.org/) is being used:
+
+```html
+<template>
+  <x-modal />
+  <router-view />
+</template>
+
+<script lang="ts" setup>
+  import XModal from "@/components/x-modal.vue";
+</script>
+```
+
+👏🏽 You should be able to see our beautiful modal rendered!
+
+## Setup the store (state management)
+
+As it's a reusable global modal, we need to keep _global_ track if this modal is opened or closed and also give the ability to control it from any other component, maintaining a unique instance, this is where Pinia comes in.
```

Comment:
```suggestion
As it's a reusable global modal, we need to keep _global_ track if this modal is opened or closed and also have the ability to control it from any other component, maintaining a unique instance. This is where Pinia comes in.
```

Reply:
El `give` solo me hace un poco de ruido. Le pondria `give it`, pero veo que usas `it` de nuevo despues. Si estamos hablando de lo mismo, pondria `...give it the ability to be controlled from any other component`